src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt

```diff
@@ -167,20 +170,106 @@ class ProjectService(
+
+        validateProjectUpdate(currentStatus, requestedStatus, fieldMask)
+
+        val finalStatus = determineFinalStatus(projectId, requestedStatus)
+        val finalRequest = request.copy {
+            this.project = request.project.copy {
+                this.status = finalStatus
+            }
         }
-        if (request.project.status == ProjectStatus.PROJECT_STATUS_DELETED) {
-            throw FailedPreconditionException("The project status can not be set to deleted by an update call.")
+
+        repo.updateProject(finalRequest).toGrpcProject()
+    }
+
+    /**
+     * Validates the update of a project based on the current status and the requested status.
+     *
+     * If the project is `PROJECT_STATUS_DELETED`, nothing can be updated.
+     * If the project is `PROJECT_STATUS_ARCHIVED`, only the status can be updated (back to an active project).
+     * If the project is `PROJECT_STATUS_ACTIVE`, then everything can be updated.
+     * If the project is `PROJECT_STATUS_ACTIVE_LOCKED`, then everything except the slr settings can be updated.
+     *
+     * In addition, the status can never be set to `PROJECT_STATUS_DELETED` via the update method.
+     *
+     * @throws FailedPreconditionException if the update fails for any reason.
+     */
+    private fun validateProjectUpdate(
+        currentStatus: ProjectStatus,
+        requestedStatus: ProjectStatus,
+        fieldMask: List<String>,
+    ) {
+        require(!(fieldMask.contains("project.status") && requestedStatus == ProjectStatus.PROJECT_STATUS_DELETED)) {
+            "The project status cannot be set to DELETED via the update method. Use SoftDeleteProject instead."
         }
 
-        repo.updateProject(request, project.status).toGrpcProject()
+        when (currentStatus) {
+            ProjectStatus.PROJECT_STATUS_DELETED -> {
+                throw FailedPreconditionException("The update failed because the project has been deleted.")
+            }
+
+            ProjectStatus.PROJECT_STATUS_ARCHIVED -> {
+                val isOnlyStatusUpdate = fieldMask.size == 1 && fieldMask.contains("project.status")
+                if (!isOnlyStatusUpdate) {
+                    throw FailedPreconditionException(
+                        "For an archived project, only the 'status' field can be updated.",
+                    )
+                }
+                require(
+                    requestedStatus == ProjectStatus.PROJECT_STATUS_ACTIVE ||
+                        requestedStatus == ProjectStatus.PROJECT_STATUS_ACTIVE_LOCKED,
+                ) { "An archived project can only be unarchived by setting its status to ACTIVE or ACTIVE_LOCKED." }
+            }
+
+            ProjectStatus.PROJECT_STATUS_ACTIVE_LOCKED -> {
+                val isChangingSettings = fieldMask.any { it.startsWith("project.settings.") }
+                require(!isChangingSettings) {
+                    "The project's settings cannot be modified while it is locked."
```

Comment:
```suggestion
                    "The project is locked and therefore no project settings can be modified."
```
But I thought that only SLR settings cannot be changed? The fetchers should be changeable, right? And the snowballing type?

Reply:
You're right. But what exactly are the SLR settings?

Reply:
After asking in the devs channel, all current settings are slr settings.

Reply:
Yeah, and your updated comment and error message are better now, imo.